### PR TITLE
CDK: assert >0 state messages per read (fix tests)

### PR DESCRIPTION
**Pull request by erohmensing**
## What
* After https://github.com/airbytehq/airbyte/pull/35905, we should be emitting a state message with every successful sync. However there are a few tests that were too lenient and weren't actually _successful_ syncs. This PR fixes those cases and adds validation that we emit at least one state message per successful sync. 

## How
* Add an assertion that we get at least 1 state message for a successful sync 
* Fix some tests that previously "output 0 expected records" but actually errored silently - do not run them as read tests
* Fix a test that failed silently due to lack of support for multi-format
* Add a new test for syncs that output 0 records successfully

## 🚨 User Impact 🚨
None - test changes

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/resources/pull-requests-handbook)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:generateScaffolds` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

**Comment by erohmensing** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/35906?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#35907** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/35907?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#35906** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/35906?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#35905** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/35905?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @erohmensing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py [hunk 1]
```diff
@@ -105,6 +105,7 @@ def _verify_read_output(output: EntrypointOutput, scenario: TestScenario[Abstrac
 
     expected_states = list(filter(lambda e: "data" not in e, expected_records))
     states = list(filter(lambda r: r.state, records))
+    assert len(states) != 0, "No state messages were emitted."
```
Comment (brianjlai) [inline airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py:108]: nit can we make this error: "No state messages emitted. Successful syncs should emit at least one stream state."

**Review by brianjlai — APPROVED** [review]
lgtm!

File: airbyte-cdk/python/unit_tests/sources/file_based/scenarios/csv_scenarios.py [hunk 2]
```diff
@@ -467,48 +467,46 @@
     )
 ).build()
 
-multi_format_analytics_scenario: TestScenario[InMemoryFilesSource] = (
+csv_analytics_scenario: TestScenario[InMemoryFilesSource] = (
```
Comment (erohmensing) [inline airbyte-cdk/python/unit_tests/sources/file_based/scenarios/csv_scenarios.py:470]: This test used to test analytics for multi-format. However, we don't actually support multi format. Although the analytics come out, the read fails silently, which means that we don't get the state message we expect. 

Instead, we test the analytics separately (for the cases we do support - one format) and don't allow a read to silently fail.

File: airbyte-cdk/python/unit_tests/sources/file_based/scenarios/csv_scenarios.py [hunk 3]
```diff
@@ -1450,7 +1487,6 @@
         }
     )
     .set_expected_discover_error(AirbyteTracedException, FileBasedSourceError.SCHEMA_INFERENCE_ERROR.value)
-    .set_expected_records([])
```
Comment (erohmensing) [inline airbyte-cdk/python/unit_tests/sources/file_based/scenarios/csv_scenarios.py:1453]: These 2 scenarios were scenarios that fail discover. They previously had "0 expected records", but fail silently if sent through the read tests. Instead, they are still discover scenarios, but no longer read scenarios - if discover fails, we don't expect the read to succeed. 

These in particular don't assert any specific expected errors, which hints toward silent failure. If we got these cases to yell more aggressively about their errors, we could bring them back as read tests with expected errors.

File: airbyte-cdk/python/unit_tests/sources/file_based/scenarios/csv_scenarios.py [hunk 4]
```diff
@@ -3009,6 +3045,61 @@
             ]
         }
     )
-    .set_expected_records([])
     .set_expected_discover_error(AirbyteTracedException, FileBasedSourceError.SCHEMA_INFERENCE_ERROR.value)
 ).build()
+
+csv_no_records_scenario: TestScenario[InMemoryFilesSource] = (
```
Comment (erohmensing) [inline airbyte-cdk/python/unit_tests/sources/file_based/scenarios/csv_scenarios.py:3051]: New scenario - a basic test for if we get 0 records in an expected, okay way. Essentially - if there's no data available, we expect 0 records, and I added this case to make sure that even if 0 records are synced, we get a state message.

File: airbyte-cdk/python/unit_tests/sources/file_based/test_file_based_scenarios.py [hunk 5]
```diff
@@ -152,7 +153,13 @@
...
 from unit_tests.sources.file_based.test_scenarios import verify_check, verify_discover, verify_read, verify_spec
 
-discover_scenarios = [
+discover_failure_scenarios = [
+    earlier_csv_scenario,
+    empty_schema_inference_scenario,
+]
+
+discover_success_scenarios = [
+    csv_no_records_scenario,
```
Comment (erohmensing) [inline airbyte-cdk/python/unit_tests/sources/file_based/test_file_based_scenarios.py:162]: See other comment - not all discover scenarios can be used as read scenarios, as some that fail discover do not read successfully (in that they don't error successfully either)

File: airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py (hunk 1 above)
Comment (erohmensing) [reply airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py:108]:
```suggestion
    assert len(states) > 0, "No state messages emitted. Successful syncs should emit at least one stream state."
```

File: airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py [hunk 6]
```diff
@@ -127,6 +128,7 @@ def _verify_read_output(output: EntrypointOutput, scenario: TestScenario[Abstrac
 
 def _verify_analytics(analytics: List[AirbyteMessage], expected_analytics: Optional[List[AirbyteAnalyticsTraceMessage]]) -> None:
     if expected_analytics:
+        assert len(analytics) == len(expected_analytics), f"Number of actual analytics messages ({len(analytics)}) did not match expected ({expected_analytics})"
```
Comment (brianjlai) [inline airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py:131]:
```suggestion
        assert len(analytics) == len(expected_analytics), f"Number of actual analytics messages ({len(analytics)}) did not match expected ({len(expected_analytics)})"
```

i forgot to add len()

**Review by brianjlai — APPROVED** [review]
✔️

File: airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py (hunk 6 above)
Comment (erohmensing) [reply airbyte-cdk/python/unit_tests/sources/file_based/test_scenarios.py]: i gotchu [77e7567](https://github.com/airbytehq/airbyte/pull/35906/commits/77e756797de61e7e1454c04c2302af5bd9656dd9)

**Comment by erohmensing** [comment]
### Merge activity

* **Mar 8, 5:21 PM EST**: @erohmensing merged this pull request with [Graphite](https://app.graphite.dev/github/pr/airbytehq/airbyte/35906).

**Pull request merged**